Gardener Community Roles

timuthy opened the pull request:
**What this PR does / why we need it**:
This PR introduces the definition of Gardener community roles, outlining the various levels and member statuses relevant to code contribution and ownership.

Gardener sub-projects may adopt this framework with technical support from Gardener’s Prow instance, aiming to establish a unified process for all contributors interested in joining the Gardener community.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
/cc @gardener/gardener-maintainers @gardener/gardener-extensions-maintainers @timebertt @dimityrmirchev @vpnachev

PR remains in draft status until the `gardener/org` repository is prepared.

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy|action
- target_group:   user|operator|developer
-->
```noteworthy developer
NONE
```

timebertt commented on website/documentation/contribute/code/roles/assets/_issue_template.md:
```diff
@@ -0,0 +1,36 @@
+---
+name: Membership Request
+about: Request a membership status in the project
```
I'm wondering if we should add this issue template only once in a central repository (maybe this one) instead of copying it to individual repositories – and then open the membership requests in this central repository.
We could add a field to the template for specifying the relevant projects. This would prevent outdated templates and reduce effort when adjusting the process.

timuthy replied:
Good suggestion! `gardener/documentation` might not be the perfect place for such requests but we can create a new repo `gardener/org` which we can use for further purposes (see also [this comment](https://github.com/gardener/documentation/pull/715#discussion_r2325113495)).

timebertt commented on website/documentation/contribute/code/roles/assets/_issue_template.md:
```diff
@@ -0,0 +1,36 @@
+---
+name: Membership Request
+about: Request a membership status in the project
```
In Kubernetes, there is the https://github.com/kubernetes/org repo, that also automates GitHub org/team membership in a GitOps-style with code reviews using Prow tooling ([Peribolos](https://docs.prow.k8s.io/docs/components/cli-tools/peribolos/)).
We could consider doing something similar to reduce manual and non-auditable actions.

timuthy replied:
I like this idea and concept in general - thanks for sharing. Nevertheless, this doesn't have the highest priority at the moment and I would rather consider it as an enhancement than a prerequisite for the framework described here.

What we can and should do first, is to create a new repo, `gardener/org`, where we store the membership request template (see [comment](https://github.com/gardener/documentation/pull/715#discussion_r2321011349)) and collect enhancements like this one.

Would you agree?

timebertt replied:
> I would rather consider it as an enhancement than a prerequisite for the framework described here.

Definitely!

> What we can and should do first, is to create a new repo, `gardener/org`, where we store the membership request template (see [comment](https://github.com/gardener/documentation/pull/715#discussion_r2321011349)) and collect enhancements like this one.

I like it, sounds like a good starting point for implementing the proposed process 👍

timebertt reviewed (commented):
Very nice, thank you for setting this up!
This definitely fills a gap in our current community setup – appreciated.

No concerns from my side, looks good in general :)

ScheererJ commented on website/documentation/contribute/code/roles/_index.md:
```diff
@@ -0,0 +1,80 @@
+
+## Community Roles
+
+This document describes contribution roles within the Gardener community, outlining their requirements and responsibilities. It aims to help new contributors understand the review process and provides guidance for those interested in gaining membership. Various aspects and definitions were adopted from the Kubernetes Community membership document ([ref](https://github.com/kubernetes/community/blo ...
```
Nit: Unnecessary space.

```suggestion
This document describes contribution roles within the Gardener community, outlining their requirements and responsibilities. It aims to help new contributors understand the review process and provides guidance for those interested in gaining membership. Various aspects and definitions were adopted from the Kubernetes Community membership document ([ref](https://github.com/kubernetes/community/blob/master/community-membership.md)), with slight adjustments to fit the organizational setup and size of Gardener.
```

ScheererJ commented on website/documentation/contribute/code/roles/_index.md:
```diff
@@ -0,0 +1,80 @@
+
+Reviewers are responsible for assessing code quality and correctness within specific areas of the project.
+They are granted the privilege to `/lgtm` pull requests.
+Members of this role are technically defined in the `OWNERS_ALIASES` file, placed at the root level of the repository.
```
As we have `Members` defined above, it might make sense to not use it in this context.

```suggestion
Contributors with this role are technically defined in the `OWNERS_ALIASES` file, placed at the root level of the repository.
```

ScheererJ commented on website/documentation/contribute/code/roles/_index.md:
```diff
@@ -0,0 +1,80 @@
+
+**Requirements**
+- Solid understanding of the codebase and software engineering principles
+- Professional expertise in at least one Gardener topic area: `core`, `networking`, `observability`, `scalability`, `security`, etc.
```
Not sure if this should be exhaustive, but if so we might want to add `infrastructure` to the list.

ScheererJ commented on website/documentation/contribute/code/roles/_index.md:
```diff
@@ -0,0 +1,80 @@
+
+**How to become an approver**
+- Serve as a reviewer for at least 3 months
+- **Process**: Obtain sponsorship from a quorum of approvers; nomination happens by one or more existing approvers
```
How is the quorum noted? Should that also be part of a PR? Maybe we should also add this similarly to the process description for the reviewer.

ScheererJ reviewed (commented):
Thanks for taking the time to formalise the process. It looks good overall.

marc1404 reviewed (commented):
Thanks for writing up the result of our discussion!
I appreciate that we're staying aligned with Kubernetes' approach to community management while adjusting it to suit our size and needs. Having this written down and being able to refer to it is an improvement.

I've left some optional formatting comments.

tobschli commented on website/documentation/contribute/code/roles/_index.md:
```diff
@@ -0,0 +1,80 @@
+- Active community engagement in the form of contributions to the Gardener Community Review meeting, Slack discussions or similar
+
+**Responsibilities**
+- Responsiveness and ability to triage reviews
```
Maybe we could also add a list of one, two items what is expected in terms of review triaging.
This is perhaps difficult to define for all repositories, but just as an example to get a feel for it

tobschli reviewed (commented):
Thank you very much for this!
I think this is really valuable in terms of community building 😊

timuthy commented on website/documentation/contribute/code/roles/assets/_issue_template.md:
```diff
@@ -0,0 +1,36 @@
+Describe the motivation of this request and your background shortly - we'd like to know you :-)
+-->
+
+**Checklist**
```
Thanks! I went with "Repository", though, as it is rather an exception to raise a membership request for multiple repositories with the same subset of reviewers and proof of record. Can still be done but shouldn't be motivated by default.

timuthy commented:
Thanks for your valuable feedback folks! I've addressed all comments - PTAL!

As agreed in [this comment](https://github.com/gardener/documentation/pull/715#discussion_r2321015493), we're currently preparing a new `gardener/org` repo and will place the issue template there.

timuthy commented:
/hold
Until `gardener/org` is prepared.

timebertt commented on website/documentation/contribute/code/roles/_index.md:
```diff
@@ -6,40 +6,43 @@ weight: 20
 - Fulfill listed requirements
 - Obtain sponsorship from at least two approvers
-- **Process**: Open an issue in the affected project repository (use [issue template](https://github.com/gardener/documentation/blob/master/website/documentation/contribute/code/roles/assets/_issue_template.md); if accepted, an org owner assigns the member to the corresponding group.
+- Open a membership request issue in [gardener/org](https://github.com/gardener/org/issues/new?template=membership.md); if accepted, an approver creates a PR to update [`OWNERS_ALIASES`](https://docs.prow.k8s.io/docs/components/plugins/approve/approvers/#overview) for the requested repository.
```
Hmm, for new members, we don't need to update the `OWNERS_ALIASES` file; we only need to invite them to the org, right?

LucaBernstein reviewed (commented):
Thank you very much for formalizing the roles. Looking good from my side as well. 😃 👍

plkokanov reviewed (approved):
Thanks again for the effort and sorry for the delay!

/lgtm
